### PR TITLE
PHOENIX-4845 Support using Row Value Constructors in OFFSET clause for paging in tables where the sort order of PK columns varies

### PR DESCRIPTION
**Pull request by dbwong**
This is an initial approach PR.  While the actual offset test case is working there are a bunch of cases to handle/clean up/refactor etc.  DO NOT MERGE

### CODE REVIEW

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/QueryWithOffsetIT.java [hunk 1]
```diff
@@ -43,191 +44,448 @@
...
+                            + "\"=" + Long.toString(500);
+            conn.createStatement().execute(query);
+        }
+
     }
 
-    @Test
-    public void testMetaDataWithOffset() throws SQLException {
+    //RVC Based Offset - Tests
+    public static class RvcOffsetTests extends ParallelStatsDisabledIT {
```
Comment (twdsilva) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/QueryWithOffsetIT.java]: move to a new class

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/QueryWithOffsetIT.java [hunk 2]
```diff
@@ -43,191 +44,448 @@
...
+            try {
+                conn.createStatement().execute(failureSql);
+            } catch (Exception e) {
+                System.out.println(e.getMessage());
+                return;
+            }
+            fail("Should not have allowed more than pk columns in Key RVC Offset");
+        }
+
+        //I think normal filtering where clauses should be reasonable but
```
Comment (twdsilva) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/QueryWithOffsetIT.java]: not sure I understand, can you provide an example of this?

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/QueryWithOffsetIT.java [hunk 3]
```diff
@@ -43,191 +44,448 @@
...
+                        ps.setInt(4,nRows);
+                        int result = ps.executeUpdate();
+                        assertEquals(1,result);
+                        nRows++;
+                    }
+                }
+            }
+            conn.commit();
+
+            //TODO: The Where clause here is currently a hack so that it goes into WhereOptimizer.pushRowKeyToScan method need to when the offset is present always go through that path
```
Comment (twdsilva) [inline phoenix-core/src/it/java/org/apache/phoenix/end2end/QueryWithOffsetIT.java]: Do you man we need to specify a where clause to get the RVC offset to execute?

**Review by twdsilva — COMMENTED** [review]
Overall approach seems better to me compared to the query rewrite approach.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/QueryWithOffsetIT.java (hunk 3 above)
Comment (dbwong) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/QueryWithOffsetIT.java]: Right now in the code would only go into pushRowKeyToScan if the where clause is not null.  I intend to put in logic to go through that path without a where clause if a rvc offset is present.  I’m undecided on the best approach to do so.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/QueryWithOffsetIT.java (hunk 2 above)
Comment (dbwong) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/QueryWithOffsetIT.java]: There are a few components to this thought.  If we fully specify a row or rows but the offset does not cover them do we allow this?  And does it make sense to have an offset in a pointget case?  The second is internal to Phoenix on how the ScanRanges class data structures usage greatly differs between point gets and range queries and some work will have to be done to handle it.

File: phoenix-core/src/it/java/org/apache/phoenix/end2end/QueryWithOffsetIT.java (hunk 1 above)
Comment (dbwong) [reply phoenix-core/src/it/java/org/apache/phoenix/end2end/QueryWithOffsetIT.java]: Will do.

**Pull request closed without merging**